File: packages/go/protocol/wire_golden_test.go
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 DeRec Alliance. All rights reserved.

package protocol

import (
	"encoding/json"
	"os"
	"reflect"
	"strings"
	"testing"
)

// goldenJSON loads one entry from the shared golden fixture
// (library/tests/fixtures/wire_golden.json), which every SDK asserts its own
// params builder against. Any difference from what that fixture pins means
// the wire format changed — do not update the fixture to make a failing
// assertion pass; fix the marshal code instead.
func goldenJSON(t *testing.T, key string) string {
	t.Helper()
	raw, err := os.ReadFile("../../../library/tests/fixtures/wire_golden.json")
	if err != nil {
		t.Fatalf("read wire_golden.json: %v", err)
	}
	var fixture map[string]json.RawMessage
	if err := json.Unmarshal(raw, &fixture); err != nil {
		t.Fatalf("parse wire_golden.json: %v", err)
	}
	value, ok := fixture[key]
	if !ok {
		t.Fatalf("wire_golden.json has no %q entry", key)
	}
	return string(value)
}

// assertJSONEqual compares two JSON documents by parsed value rather than
// raw bytes: the fixture is pretty-printed while the marshaled output under
// test is compact, so a byte comparison would always fail even when the
// values are identical.
func assertJSONEqual(t *testing.T, got, want string) {
	t.Helper()
	var gotValue, wantValue map[string]any
	if err := json.Unmarshal([]byte(got), &gotValue); err != nil {
		t.Fatalf("parse got JSON: %v", err)
	}
	if err := json.Unmarshal([]byte(want), &wantValue); err != nil {
		t.Fatalf("parse want JSON: %v", err)
	}
	if !reflect.DeepEqual(gotValue, wantValue) {
		t.Fatalf("wire format changed:\n got  = %s\n want = %s", got, want)
	}
}

// goldenRestoreSecret builds the same fully-populated Secret the golden
// bytes above were captured from: two helpers (one with
// CommunicationInfo, one without — exercising the omitempty branch both
// ways), two UserSecrets, and a populated Replicas with two entries (same
// CommunicationInfo-present/absent split).
func goldenRestoreSecret() Secret {
	sharedKey1 := make([]byte, 32)
	for i := range sharedKey1 {
		sharedKey1[i] = byte(i)
	}
	sharedKey2 := make([]byte, 32)
	for i := range sharedKey2 {
		sharedKey2[i] = byte(31 - i)
	}
	replicaGroupKey := []byte{9, 8, 7, 6, 5, 4, 3, 2, 1, 0}

	return Secret{
		Helpers: []Helper{
			{
				ChannelID:         "11",
				Transports:        []EndpointJSON{{URI: "https://helper-a.example.com", Protocol: 0}},
				SharedKey:         sharedKey1,
				CommunicationInfo: map[string]string{"foo": "bar"},
			},
			{
				ChannelID:    "22",
				Transports:        []EndpointJSON{{URI: "https://helper-b.example.com", Protocol: 0}},
				SharedKey:    sharedKey2,
			},
		},
		Secrets: []UserSecret{
			{ID: []byte{0x01}, Name: "wallet", Data: []byte("correct horse battery staple")},
			{ID: []byte{0x02, 0x03}, Name: "seed", Data: []byte{0xde, 0xad, 0xbe, 0xef}},
		},
		Replicas: &Replicas{
			ChannelID: "33",
			Members: []Replica{
				{
					ReplicaID:         "44",
					Transports:        []EndpointJSON{{URI: "https://replica-a.example.com", Protocol: 0}},
					Role:              "Source",
					CommunicationInfo: map[string]string{"baz": "qux"},
				},
				{
					ReplicaID:    "66",
					Transports:        []EndpointJSON{{URI: "https://replica-b.example.com", Protocol: 0}},
					Role:         "Destination",
				},
			},
			SharedKey: replicaGroupKey,
		},
	}
}

// TestWireGolden_Restore asserts Restore's params marshal (via
// restoreParamsWire, whose RecoveredSecret field is now a plain Secret —
// see flow.go) matches the shared golden fixture that every SDK's params
// builder is checked against.
func TestWireGolden_Restore(t *testing.T) {
	got, err := json.Marshal(restoreParamsWire{
		Version:         7,
		RecoveredSecret: goldenRestoreSecret(),
	})
	if err != nil {
		t.Fatalf("marshal restore params: %v", err)
	}
	assertJSONEqual(t, string(got), goldenJSON(t, "restore"))
}

// TestWireGolden_ProtectSecret asserts ProtectSecretParams' marshal (via
// marshalFlowParams, whose protectSecretParamsWire.Secrets field is now
// []UserSecret directly — see flow.go) matches the shared golden fixture
// that every SDK's params builder is checked against.
func TestWireGolden_ProtectSecret(t *testing.T) {
	secret := goldenRestoreSecret()
	desc := "capture description"
	got, err := marshalFlowParams(FlowKindProtectSecret, ProtectSecretParams{
		Secrets:     secret.Secrets,
		Description: &desc,
	})
	if err != nil {
		t.Fatalf("marshal protect secret params: %v", err)
	}
	assertJSONEqual(t, string(got), goldenJSON(t, "protect_secret"))
}

// TestWireGolden_Restore_NilReplicasOmitsField pins the other end of the
// omitempty behavior: a Secret with no Replicas setup must omit the
// "replicas" key entirely, matching wire.rs's
// `#[serde(skip_serializing_if = "Option::is_none")]` on SecretWire.replicas
// — not marshal it as a JSON null (which secret.MarshalJSON's plain
// `*Replicas` field, without the custom method, would have produced).
func TestWireGolden_Restore_NilReplicasOmitsField(t *testing.T) {
	secret := Secret{
		Helpers: []Helper{
			{ChannelID: "1", Transports:        []EndpointJSON{{URI: "https://h.example.com", Protocol: 0}}, SharedKey: make([]byte, 32)},
		},
		Secrets: []UserSecret{{ID: []byte{0x01}, Name: "n", Data: []byte{0x02}}},
	}
	got, err := json.Marshal(restoreParamsWire{Version: 1, RecoveredSecret: secret})
	if err != nil {
		t.Fatalf("marshal restore params: %v", err)
	}
	if strings.Contains(string(got), `"replicas"`) {
		t.Fatalf("expected no replicas key for a nil Replicas pointer, got: %s", got)
	}
}

File: packages/react-native/__tests__/wire-golden.test.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 DeRec Alliance. All rights reserved.

import * as fs from 'node:fs';
import * as path from 'node:path';

import { buildProtectSecretParams, buildRestoreParams } from '../src/protocol';
import type { DeRecEvent, ProtectSecretParams } from '../src/types';

const golden = JSON.parse(
  fs.readFileSync(path.join(__dirname, '../../../library/tests/fixtures/wire_golden.json'), 'utf8'),
) as Record<string, unknown>;

function asciiBytes(text: string): Uint8Array {
  return Uint8Array.from(Array.from(text, (c) => c.charCodeAt(0)));
}

// Mirrors goldenRestoreSecret() at packages/go/protocol/wire_golden_test.go:59:
// two UserSecrets shared by both the restore and protect_secret fixtures.
function goldenUserSecrets(): ProtectSecretParams['secrets'] {
  return [
    { id: Uint8Array.from([0x01]), name: 'wallet', data: asciiBytes('correct horse battery staple') },
    { id: Uint8Array.from([0x02, 0x03]), name: 'seed', data: Uint8Array.from([0xde, 0xad, 0xbe, 0xef]) },
  ];
}

function goldenProtectSecretInput(): ProtectSecretParams {
  return {
    secrets: goldenUserSecrets(),
    description: 'capture description',
  };
}

// The two 32-byte shared keys are filled with `i` and `31 - i` respectively,
// matching goldenRestoreSecret()'s sharedKey1/sharedKey2. helpers[1] and
// replicas.members[1] carry an empty communication_info map (the field is
// required on this event shape) to exercise the same omit-when-empty branch
// Go's `omitempty` exercises on its CommunicationInfo-absent entries.
function goldenRestoreInput(): Extract<DeRecEvent, { type: 'SecretRecovered' }>['secret'] {
  const sharedKey1 = Uint8Array.from({ length: 32 }, (_, i) => i);
  const sharedKey2 = Uint8Array.from({ length: 32 }, (_, i) => 31 - i);
  const replicaGroupKey = Uint8Array.from([9, 8, 7, 6, 5, 4, 3, 2, 1, 0]);

  return {
    helpers: [
      {
        channel_id: '11',
        transport_uri: 'https://helper-a.example.com',
        shared_key: sharedKey1,
        communication_info: { foo: 'bar' },
      },
      {
        channel_id: '22',
        transport_uri: 'https://helper-b.example.com',
        shared_key: sharedKey2,
        communication_info: {},
      },
    ],
    secrets: goldenUserSecrets(),
    replicas: {
      channel_id: '33',
      members: [
        {
          replica_id: '44',
          transport_uri: 'https://replica-a.example.com',
          role: 'Source',
          communication_info: { baz: 'qux' },
        },
        {
          replica_id: '66',
          transport_uri: 'https://replica-b.example.com',
          role: 'Destination',
          communication_info: {},
        },
      ],
      shared_key: replicaGroupKey,
    },
  };
}

describe('wire golden vectors', () => {
  it('protect_secret params match the shared fixture', () => {
    expect(buildProtectSecretParams(goldenProtectSecretInput())).toEqual(golden.protect_secret);
  });

  it('restore params match the shared fixture', () => {
    expect(buildRestoreParams(goldenRestoreInput(), 7)).toEqual(golden.restore);
  });
});
